docs/concepts/function-components/custom-hooks.md

```diff
@@ -1,12 +1,105 @@
 ---
-title: Custom hooks
-description: Defining your own hooks 
+title: Custom Hooks
+description: Defining your own Hooks 
 ---
 
-## Defining custom hooks
+## Defining custom Hooks
 
-:::important contribute
-Contribute to our docs: Add details about defining custom hooks
+Component's stateful logic can be extracted into usable function by creating custom Gooks. 
 
-Take a look at the [project board](https://github.com/yewstack/yew/projects/3) for details and consider helping out.
-::: 
+Consider that we have a component which subscribes to an agent and displays the messages sent to it.
+```rust
+#[function_component(ShowMessages)]
+pub fn show_messages() -> Html {
+    let (state, set_state) = use_state(|| RefCell::new(vec![]));
```

Comment:
Unfortunately, that is impossible. The `Vec` inside `Rc` needs to be mutated to add the new message before passing it to `set_state` which can't happen as explained in [this SO thread](https://stackoverflow.com/questions/58599539/cannot-borrow-in-a-rc-as-mutable).

Reply:
That's true, but we don't need interior mutability here because you have to call `set_state` with an owned value anyway.
Currently you're doing that using `(*state).clone()` which dereferences `Rc` to `&Vec<_>` and then clones it, making it `Vec<_>`.
So you're using `borrow_mut` to push an item to the vector but then it's immediately replaced by a clone of this value.

This is how easy it could be:
```rust
EventBus::bridge(Callback::from(move |msg| {
    let mut messages = (*state).clone();
    messages.push(msg);
    set_state(messages)
}));
```